osutil/env.go

```diff
@@ -50,3 +53,56 @@ func GetenvBool(key string, dflt ...bool) bool {
 
 	return b
 }
+
+// SubstituteEnv takes a list of environment strings like:
+// - K1=BAR
+// - K2=$BAR
+// - K3=${BAZ}
+// and substitutes them using the os environment strings.
+//
+// Strings that do not have the form "k=v" will be dropped.
+//
+// The result will be sorted list of environment strings.
+//
+// Circular references (A=$B,B=$A) will result in empty A,B
+// (just like in shell)
+func SubstituteEnv(env []string) []string {
+	envMap := map[string]string{}
+	for _, s := range env {
+		l := strings.SplitN(s, "=", 2)
+		if len(l) < 2 {
+			continue
+		}
+		envMap[l[0]] = l[1]
+	}
+
+	// this always terminates, in each iteration of the loop we
+	// eliminate at least one value with a "$var"
```

Comment:
Note that at the moment something defined in the outer-most level may refer to a variable that is in the innermost scope. I started with that same logic in spread and regretted for a few reasons. Ended up going down to a more sequential and shell-like logic where what comes first may be used later, even in the local scope. This forces some basic organization, which ends up also making it more readable for people to parse.

[Here](https://github.com/snapcore/spread/blob/master/spread/project.go#L186) is the piece that reads the yaml preserving the map ordering, and ensuring some basic consistency (no double keys, for example).